sirepo/pkcli/runner.py

```diff
@@ -105,122 +93,192 @@ def _write_status(status, run_dir):
-                # environment, so we have to manually switch back to py2 mode.
-                env['PYENV_VERSION'] = 'py2'
-                cmd = ['pyenv', 'exec'] + cmd
-                with open(run_log_path, 'a+b') as run_log:
-                    process = trio.Process(
-                        cmd,
-                        cwd=run_dir,
-                        start_new_session=True,
-                        stdin=subprocess.DEVNULL,
-                        stdout=run_log,
-                        stderr=run_log,
-                        env=env,
-                    )
-                self.jobs[run_dir] = _JobInfo(
-                    run_dir, jhash, runner_client.JobStatus.RUNNING, process
+            else:
+                # It's some other job. Better kill it before doing
+                # anything else.
+                # XX TODO: should we check some kind of sequence number
+                # here? I don't know how those work.
+                pkdlog(
+                    '{} {}: stale job is still running; killing it',
+                    run_dir, jhash,
                 )
-                async with process:
-                    task_status.started()
-                    # XX more race conditions here, in case we're writing to
-                    # the wrong version of the directory...
-                    await process.wait()
-                    if process.returncode:
+                await self.kill_all(run_dir)
+
+        # Okay, now we have the dir to ourselves. Set up the new run_dir:
+        assert run_dir not in self.report_jobs
+        pkio.unchecked_remove(run_dir)
+        tmp_dir.rename(run_dir)
+        # Start the job:
+        report_job = await _BACKENDS[backend].start_report_job(run_dir, cmd)
+        # And update our records so we know it's running:
+        job_info = _JobInfo(
+            run_dir, jhash, runner_client.JobStatus.RUNNING, report_job,
+        )
+        self.report_jobs[run_dir] = job_info
+        pkjson.dump_pretty(
+            {
+                'backend': backend,
+                'info': report_job.backend_info,
+            },
+            filename=run_dir.join(_BACKEND_INFO_BASENAME),
+        )
+
+        # And finally, start a background task to watch over it.
+        self._nursery.start_soon(
+            self._supervise_report_job, run_dir, jhash, job_info,
+        )
+
+    async def _supervise_report_job(self, run_dir, jhash, job_info):
+        with _catch_and_log_errors(Exception, 'error in _supervise_report_job'):
+            # Make sure returncode is defined in the finally block, even if
+            # wait() somehow crashes
+            returncode = None
+            try:
+                returncode = await job_info.report_job.wait()
+            finally:
+                async with self.locks[run_dir]:
+                    # Clear up our in-memory status
+                    assert self.report_jobs[run_dir] is job_info
```

Comment:
Eh, I guess `is` is probably a tiny smidgen faster. But really I used `is` because conceptually, what we want to check here is that no-one's replaced the `job_info` object behind our back, no matter how similar it might or might not be. In this case it makes no difference, because `==` on `_JobInfo` uses `object.__eq__` which does an `is` comparison anyway, but if we did have an `__eq__` override then I think it would still be correct to use `is` here.